lib/Sema/CSSimplify.cpp

```diff
@@ -5275,6 +5275,19 @@ bool ConstraintSystem::repairFailures(
     return true;
   }
 
+  auto maybeRepairKeyPathResultFailure = [&](KeyPathExpr *kpExpr) {
+      if (isKnownKeyPathType(lhs) && isKnownKeyPathType(rhs)) {
+        // If we have keypath capabilities for both sides and one of the bases
+        // is unresolved, it is too early to record fix.
+        if (hasConversionOrRestriction(ConversionRestrictionKind::DeepEquality))
```

Comment:
The possible failure modes are:

- key path types mismatch on capability - KeyPath vs. WritableKeyPath for example
- one of the generic parameters (either Root or Value) mismatch:
  - If there is any conversion recorded against this match we should let that happen (not just deep equality)
  - otherwise we could either just record whole types or try to deep deeper and recursively call repairFailures on Root and Value.

Reply:
@xedin This bit I had just pulled up from lines 5296-5305 below for reuse. Happy to update it but not sure I totally understand the restricted conversion machinery. From your bullets it sounds like you're saying we should attempt the fix unconditionally if we have a capability mismatch, but if they agree on capability then we should only attempt the fix if we have *no* restrictions, does that sound right? So that would amount to updating the check to something like the following:

```suggestion
        if (lhs->getAnyGeneric() == rhs->getAnyGeneric() && llvm::any_of(conversionsOrFixes,
                        [](const RestrictionOrFix &correction) {
                          return bool(correction.getRestriction());
                        }))
```
(seems like this may be the best way to check keypath capability from just a `Type` instance, unless I'm missing something?)

Reply:
I'd say:

```
if (lhs->getAnyGeneric() != rhs->getAnyGeneric() || llvm::any_of(conversionsOrFixes, ...))
```

Because if key paths have different capatibilities we don't want to diagnose it here, same goes for restrictions - each restriction would diagnose separately when applied.

Unfortunate consequence of the current state of things is that every key path conversion anchored on KeyPathExpr ends up forming a disjunction where second choice has a fix and we'd really match try to avoid that.

Reply:
Hm, it appears there are some circumstances where this is the *only* place we end up applying a fix for mismatched capabilities, e.g.:

```
func issue_65965() {
  struct S {
	  var s: String
	  let v: String
  }
	
  let refKP: ReferenceWritableKeyPath<S, String>
  refKP = \.s
  // expected-error@-1 {{key path value type 'WritableKeyPath<S, String>' cannot be converted to contextual type 'ReferenceWritableKeyPath<S, String>'}}
}
```

If we don't apply the fix here then we end up with the general 'type of expression is ambiguous without type annotation' fallback error as opposed to the specific error. Where would you have expected this to get diagnosed if not here?

Reply:
Looks like what happens is that 1) we infer `ReferenceWritableKeyPath` as the type for the literal via the assignment operator, and then in `simplifyKeyPathConstraint` attempts to bind to that type rather than convert, so we never have a chance to apply conversion restriction, and 2) even if we change things to allow conversion here since the mismatch ends up happening directly at the keypath constraint we don't even qualify for applying the fix in `simplifyRestrictedConstraint`:
```
if (loc->isForAssignment() || loc->isForCoercion() ||
        loc->isForContextualType() ||
        loc->isLastElement<LocatorPathElt::ApplyArgToParam>() ||
        loc->isForOptionalTry()) {
```

Reply:
Hm, we can leave capability out of this for now, I think the only way to fix it would be to always let `simplifyKeyPathConstraint` assign a contextual type instead of eagerly binding it, that way it'd always end up with a mismatch in the right place.